src/awkward/operations/convert.py

```diff
@@ -3367,23 +3292,387 @@ def __call__(self, row_group, unpack, length, form, lazy_cache, lazy_cache_key):
+            "lst:"
+        )
+        samplekey = "{0}:off:{1}:{2}[{3}]".format(
+            lazy_cache_key,
+            sampleform.form_key[4:],
+            ".".join("" if x is None else x for x in columns),
+            row_group,
+        )
+        sample = None
+        if lazy_cache is not None:
+            try:
+                sample = lazy_cache.mutablemapping[samplekey]
+            except KeyError:
+                pass
+        if sample is None:
+            sample = self.get(row_group, columns, sampleform, struct_only)
+        if lazy_cache is not None:
+            lazy_cache.mutablemapping[samplekey] = sample
+
+        offsets = [sample.offsets]
+        sublength = offsets[-1][-1]
+        sample = sample.content
+        recordform = form.content
+        columns = columns + (None,)
+        while not isinstance(recordform, ak.forms.RecordForm):
+            offsets.append(sample.offsets)
+            sublength = offsets[-1][-1]
+            sample = sample.content
+            recordform = recordform.content
+            columns = columns + (None,)
+
+        out = self(
+            row_group, columns, sublength, recordform, lazy_cache, lazy_cache_key
+        )
+        for off in offsets[::-1]:
+            if isinstance(off, ak.layout.Index32):
+                out = ak.layout.ListOffsetArray32(off, out)
+            elif isinstance(off, ak.layout.Index64):
+                out = ak.layout.ListOffsetArray64(off, out)
+            else:
+                raise AssertionError(
+                    "unexpected Index type: {0}".format(off)
+                    + ak._util.exception_suffix(__file__)
+                )
+        return out
+
+    def _find_first_column(self, form, struct_only):
+        if isinstance(form, ak.forms.VirtualForm):
+            return self._find_first_column(form.form, struct_only)
+        elif isinstance(form, ak.forms.RecordForm):
+            assert form.numfields != 0
+            struct_only.insert(0, form.key(0))
+            return self._find_first_column(form.content(0), struct_only)
+        elif isinstance(form, ak.forms.ListOffsetForm):
+            if form.parameter("__array__") in ("string", "bytestring"):
+                return form
+            else:
+                return self._find_first_column(form.content, struct_only)
+        else:
+            return form
 
     def get(self, row_group, unpack, form, struct_only):
         assert form.form_key.startswith("col:") or form.form_key.startswith("lst:")
         column_name = form.form_key[4:]
         masked = isinstance(form, ak.forms.ByteMaskedForm)
         if masked:
             form = form.content
-        table = self.read(row_group, column_name)
-        return _ParquetFile_arrow_to_awkward(table, struct_only, masked, unpack)
+        table = self._reader.read_row_group(row_group, [column_name])
+        return self._convert_arrow_to_awkward(table, struct_only, masked, unpack)
+
+
+class _ArrowReader(object):
```

Comment:
As a superclass of three Parquet readers, I don't think the name "`_ArrowReader`" is appropriate. These are really all about Parquet, not Arrow. Arrow doesn't have concepts like "row groups," which is in most of the common method names. The name "`_ParquetReader`" isn't used by any of the subclasses, so that would be a good replacement.

Reply:
Oh, so Arrow is aware of row groups now. Okay, but still, "Arrow" confuses me about the intention of this class. Since you chose the name by the fact that it's part of the Dataset interface, would "`_Dataset`" be a good name?

On the scale of things, this is the bunny, while `[[]] * len(columns)` was the tiger...

<img src="https://user-images.githubusercontent.com/1852447/122956844-fa925e80-d346-11eb-93a4-860c82fe0293.png" width="400">

Reply:
Haha perfect. I'm not one to bikeshed names - let's go with dataset